### PR TITLE
Implement scroll events

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

**Comment by marijnvdwerf** [comment]
@AaronVanGeffen Could you go over the changes made in this PR so far? I do want to clean up stuff (like merging scroll 1 to 4) but I first want to make sure we're matching vanilla.

File: src/openloco/ui/scrollview.h [hunk 1]
```diff
@@ -21,6 +21,29 @@ namespace openloco::ui::scrollview
         vscrollbar_thumb = 10,
     };
 
+    namespace scroll_flags
+    {
+        constexpr uint16_t HSCROLLBAR_VISIBLE = (1 << 0);
```
Comment (AaronVanGeffen) [inline src/openloco/ui/scrollview.h]: Low priority, but these should probably be refactored into lowercase.

File: src/openloco/input/mouse_input.cpp [hunk 2]
```diff
@@ -307,6 +310,110 @@ namespace openloco::input
...
+    static void loc_4C72ED()
+    {
+        auto window = WindowManager::find(_pressedWindowType, _pressedWindowNumber);
+        if (window == nullptr)
+            return;
+
+        constexpr uint16_t horizontalFlags = scroll_flags::HSCROLLBAR_THUMB_PRESSED | scroll_flags::HSCROLLBAR_LEFT_PRESSED | scroll_flags::HSCROLLBAR_RIGHT_PRESSED;
+        constexpr uint16_t verticalFlags = scroll_flags::VSCROLLBAR_THUMB_PRESSED | scroll_flags::VSCROLLBAR_UP_PRESSED | scroll_flags::VSCROLLBAR_DOWN_PRESSED;
+
+        window->scroll_areas[_currentScrollOffset / sizeof(ui::scroll_area_t)].flags &= ~(verticalFlags | horizontalFlags);
```
Comment (AaronVanGeffen) [inline src/openloco/input/mouse_input.cpp]: This looks quite different from the assembly. Either way, I think the scroll_area index should be computed on its own line.

File: src/openloco/input/mouse_input.cpp [hunk 3]
```diff
@@ -307,6 +310,110 @@ namespace openloco::input
...
+                {
+                    loc_4C72ED();
+                    return;
+                }
+
+                if (_currentScrollArea == scrollview::scroll_part::hscrollbar_thumb)
+                {
+                    int16_t deltaX = x - _tooltipCursorX;
+                    _tooltipCursorX = x;
+                    scrollview::scroll_5_follow(_pressedWindowType, _pressedWindowNumber, _pressedWidgetIndex, _currentScrollOffset / sizeof(ui::scroll_area_t), deltaX);
```
Comment (AaronVanGeffen) [inline src/openloco/input/mouse_input.cpp]: Again, I'm not following the `_currentScrollOffset / sizeof(ui::scroll_area_t)`.

File: src/openloco/ui/scrollview.cpp [hunk 4]
```diff
@@ -2,11 +2,161 @@
...
+    static constexpr uint16_t barWidth = 11;
+
+    // 0x004C87E1
+    void scroll_5_follow(WindowType _pressedWindowType, window_number _pressedWindowNumber, widget_index _pressedWidgetIndex, size_t _currentScrollOffset, int16_t ax)
+    {
+        auto window = WindowManager::find(_pressedWindowType, _pressedWindowNumber);
+        if (window == nullptr)
+            return;
+
+        scroll_area_t& scrollArea = window->scroll_areas[_currentScrollOffset];
```
Comment (AaronVanGeffen) [inline src/openloco/ui/scrollview.cpp]: At this point, this isn't `_currentScrollOffset` anymore. You've already divided it in the call.

File: src/openloco/ui/scrollview.cpp [hunk 5]
```diff
@@ -2,11 +2,161 @@
...
+        scrollArea.flags |= scroll_flags::HSCROLLBAR_THUMB_PRESSED;
+
+        uint16_t trackWidth = w.width() - 2 - thumbSize - thumbSize;
+        if (scrollArea.flags & scroll_flags::VSCROLLBAR_VISIBLE)
+        {
+            trackWidth -= barWidth;
+        }
+
+        float scrollStepSize = (float)scrollArea.contentWidth / (float)trackWidth;
+        ax = std::round(ax * scrollStepSize);
```
Comment (AaronVanGeffen) [inline src/openloco/ui/scrollview.cpp]: I don't see the advantage over doing an integer division. Looking at https://gcc.godbolt.org/z/NoQOse, that generates an `idiv`, just like the assembly,

Comment (marijnvdwerf) [reply src/openloco/ui/scrollview.cpp]: More obvious/clear code when reading it. If it gets optimised away: all the better.

Comment (AaronVanGeffen) [reply src/openloco/ui/scrollview.cpp]: These floats are not likely to be optimised away — that's why I'm commenting here. https://gcc.godbolt.org/z/ofG5kE

File: src/openloco/ui/scrollview.cpp [hunk 6]
```diff
@@ -2,11 +2,160 @@
...
+        WindowManager::invalidateWidget(_pressedWindowType, _pressedWindowNumber, _pressedWidgetIndex);
+    }
+
+    // 0x004C8898
+    void scroll_10_follow(WindowType _pressedWindowType, window_number _pressedWindowNumber, widget_index _pressedWidgetIndex, size_t _currentScrollOffset, int16_t ax)
+    {
+    }
+
+    // 0x004C894F
+    void scroll_1(WindowType _pressedWindowType, window_number _pressedWindowNumber, widget_index _pressedWidgetIndex, size_t _currentScrollOffset)
```
Comment (AaronVanGeffen) [inline src/openloco/ui/scrollview.cpp:57]: This was also implemented by #557. Pity.

File: src/openloco/ui/scrollview.cpp [hunk 7]
```diff
@@ -2,11 +2,160 @@
...
+        maxOffset = std::max<int16_t>(maxOffset, 0);
+
+        scrollArea.contentOffsetX = std::min<int16_t>(newOffset, maxOffset);
+
+        scrollview::update_thumbs(window, _pressedWidgetIndex);
+        WindowManager::invalidateWidget(_pressedWindowType, _pressedWindowNumber, _pressedWidgetIndex);
+    }
+
+    // 0x004C8B26
+    void scroll_6(WindowType, window_number, widget_index, size_t) {}
```
Comment (duncanspumpkin) [inline src/openloco/ui/scrollview.cpp:149]: Implemented in #563

File: src/openloco/ui/scrollview.cpp [hunk 8]
```diff
@@ -2,11 +2,160 @@
...
+
+        int16_t newOffset = scrollArea.contentOffsetX - 3;
+        scrollArea.contentOffsetX = std::max<int16_t>(newOffset, 0);
+
+        scrollview::update_thumbs(window, _pressedWidgetIndex);
+        WindowManager::invalidateWidget(_pressedWindowType, _pressedWindowNumber, _pressedWidgetIndex);
+    }
+
+    // 0x004C89AE
+    void scroll_2(WindowType _pressedWindowType, window_number _pressedWindowNumber, widget_index _pressedWidgetIndex, size_t _currentScrollOffset)
```
Comment (duncanspumpkin) [inline src/openloco/ui/scrollview.cpp:75]: Implemented in #562

File: src/openloco/ui/scrollview.cpp [hunk 9]
```diff
@@ -2,11 +2,160 @@
...
+
+        scrollview::update_thumbs(window, _pressedWidgetIndex);
+        WindowManager::invalidateWidget(_pressedWindowType, _pressedWindowNumber, _pressedWidgetIndex);
+    }
+
+    // 0x004C8B26
+    void scroll_6(WindowType, window_number, widget_index, size_t) {}
+
+    // 0x004C8B85
+    void scroll_7(WindowType, window_number, widget_index, size_t) {}
```
Comment (AaronVanGeffen) [inline src/openloco/ui/scrollview.cpp:152]: Implemented by #564

File: src/openloco/ui/scrollview.cpp [hunk 10]
```diff
@@ -2,11 +2,160 @@
...
+
+    // 0x004C8C0D
+    void scroll_8(WindowType, window_number, widget_index, size_t) {}
```
Comment (AaronVanGeffen) [inline src/openloco/ui/scrollview.cpp:155]: Implemented by #566

**Comment by duncanspumpkin** [comment]
I've rebased and condensed this into two commits for the new content. https://github.com/OpenLoco/OpenLoco/compare/master...duncanspumpkin:scroll-events do you want me to overwrite this PR with that branch?

**Comment by marijnvdwerf** [comment]
Sure. Better an updated PR than a stale one :)

**Comment by duncanspumpkin** [comment]
This has now been reworked to take in the refactoring and is ready to go.

**Pull request merged**